use dbus instead of localectl

joejulian opened the pull request:
Alternative fix for #26805.

jfindlay committed:
add default param in _parse_localectl in locale mod

Fixes #26805.

Commit:
Use dbus directly

Commit:
fix indent change

jfindlay commented:
Nice, thanks @joejulian.  I don't know much about dbus, but is it guaranteed to be available on all systemd systems?

joejulian commented:
Yes it is. There's even effort being made to merge it in to the kernel. https://github.com/gregkh/kdbus

joejulian commented:
(sigh) I thought I had vim set up to not leave trailing spaces...

Commit:
remove trailing spaces

cachedout commented on salt/modules/localemod.py:
```diff
@@ -8,6 +8,7 @@
 import logging
 import re
 import os
+import dbus
```
We need to gate this import and then return False from `__virtual__` if it is not installed.

cachedout commented:
Hi @joejulian. Please see my in-line comments here. Once we get that fixed up, we can go ahead and get this merged.

Commit:
Gated dbus for os families that use it

Gated the dbus include for os families that are currently using it
within this module at this time.

Renamed some defs to more appropriately express their function.

Commit:
pep8 fixes

anlutro commented:
There seems to be lots of calls to the dbus module even if `HAS_DBUS == False` and `_uses_dbus() == False`. On Debian:

```
          ID: locale-set_system
    Function: locale.system
        Name: en_US.UTF-8
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/tmp/.root_3cd8c3_salt/py2/salt/state.py", line 1701, in call
                  **cdata['kwargs'])
                File "/tmp/.root_3cd8c3_salt/py2/salt/loader.py", line 1607, in wrapper
                  return f(*args, **kwargs)
                File "/tmp/.root_3cd8c3_salt/py2/salt/states/locale.py", line 49, in system
                  if __salt__['locale.get_locale']() == name:
                File "/tmp/.root_3cd8c3_salt/py2/salt/modules/localemod.py", line 134, in get_locale
                  return _locale_get()
                File "/tmp/.root_3cd8c3_salt/py2/salt/modules/localemod.py", line 84, in _locale_get
                  return _parse_dbus_locale().get('LANG', '')
                File "/tmp/.root_3cd8c3_salt/py2/salt/modules/localemod.py", line 61, in _parse_dbus_locale
                  bus = dbus.SystemBus()
              NameError: global name 'dbus' is not defined
```

joejulian commented on salt/modules/localemod.py:
```diff
@@ -24,38 +30,55 @@ def __virtual__():
     Only work on POSIX-like systems
     '''
+    if HAS_DBUS is False and _uses_dbus():
+        return False
```
Oops, logic fail.

anlutro commented:
I don't even know how to approach fixing this? Why is dbus the only thing we're checking for suddenly?

joejulian commented:
Looks like it's fixed in the current 2015.8 branch.

joejulian commented:
@anlutro Here's the reason for using dbus on systemd systems: https://github.com/saltstack/salt/issues/26805#issuecomment-136838444

anlutro commented:
I got a bit confused by the git logs - the breaking change is actually not related to this PR.

anlutro commented:
See https://github.com/saltstack/salt/pull/31552
